crypto/dsa/dsa_ossl.c

```diff
@@ -189,17 +198,16 @@ static int dsa_sign_setup(DSA *dsa, BN_CTX *ctx_in,
 
     /*
      * We do not want timing information to leak the length of k, so we
-     * compute g^k using an equivalent exponent of fixed length. (This
-     * is a kludge that we need because the BN_mod_exp_mont() does not
-     * let us specify the desired timing behaviour.)
+     * compute G*k using an equivalent scalar of fixed bit-length.
+     *
+     * We unconditionally perform both of these additions to prevent a
+     * small timing information leakage.  We then choose the sum that is
+     * one bit longer than the modulus.
      */
-
-    if (!BN_add(k, k, dsa->q))
+    if (!BN_add(l, k, dsa->q)
+        || !BN_add(m, l, dsa->q)
+        || !BN_copy(k, BN_num_bits(l) > BN_num_bits(dsa->q) ? l : m))
```

Comment:
As mentioned earlier, formally speaking this conditional move leaks same information as conditional addition. So one could as well `k = BN_num_bits(l) > BN_num_bits(dsa->q) ? l : m`. Well, one can argue that copying minimizes window of opportunity, but one should really aim for memory-access-agnostic conditional copy. Maybe it would be appropriate to leave a TODO note saying that a alternative BN_copy should be implemented at some point...

Reply:
I don't see why is it leaking information.  The copy shouldn't be -- it always copies the same number of bits and there is no resizing involved.  Both `l` and `m` should be cached after the `BN_add`.  What have I missed?

The comparison and ?: operator might leak something.  My thought is any variation in these two will be swamped by the other operations.  Still, I could switch them to the constant time flavours if there is a need.

Reply:
It's copying from *distinct* places, and formally speaking attacker has a shot on telling them apart. And if he can [tell them apart] it would be equivalent leakage. It's not *direct* timing attack, but attacker would still use timing information to compare access times. Memory-access-agnostic conditional copy on the other hand would always read *both* inputs and mask one of them, thus making it impossible to tell which one was actually copies to destination.

Reply:
Ah, okay I understand.  For the future then.